### ISSUE TITLE
Selenium::WebDriver::Error::InvalidArgumentError: Expected [object Undefined] undefined to be a string

### ISSUE THREAD

**Issue by khanimdad** [mozilla/geckodriver#707]
On firefox 53.0 (64bits) for Ubuntu, am getting this error, below are error trace...

```
irb(main):019:0> browser.text_field(:id => "login-email").set(USERNAME)
Selenium::WebDriver::Error::InvalidArgumentError: Expected [object Undefined] undefined to be a string
	from /var/lib/gems/2.3.0/gems/selenium-webdriver-3.3.0/lib/selenium/webdriver/remote/response.rb:69:in `assert_ok'
	from /var/lib/gems/2.3.0/gems/selenium-webdriver-3.3.0/lib/selenium/webdriver/remote/response.rb:32:in `initialize'
	from /var/lib/gems/2.3.0/gems/selenium-webdriver-3.3.0/lib/selenium/webdriver/remote/http/common.rb:83:in `new'
	from /var/lib/gems/2.3.0/gems/selenium-webdriver-3.3.0/lib/selenium/webdriver/remote/http/common.rb:83:in `create_response'
	from /var/lib/gems/2.3.0/gems/selenium-webdriver-3.3.0/lib/selenium/webdriver/remote/http/default.rb:107:in `request'
	from /var/lib/gems/2.3.0/gems/selenium-webdriver-3.3.0/lib/selenium/webdriver/remote/http/common.rb:61:in `call'
	from /var/lib/gems/2.3.0/gems/selenium-webdriver-3.3.0/lib/selenium/webdriver/remote/w3c_bridge.rb:620:in `execute'
	from /var/lib/gems/2.3.0/gems/selenium-webdriver-3.3.0/lib/selenium/webdriver/remote/w3c_bridge.rb:418:in `send_keys_to_element'
	from /var/lib/gems/2.3.0/gems/selenium-webdriver-3.3.0/lib/selenium/webdriver/common/element.rb:156:in `send_keys'
	from /var/lib/gems/2.3.0/gems/watir-6.2.0/lib/watir/user_editable.rb:13:in `block in set'
	from /var/lib/gems/2.3.0/gems/watir-6.2.0/lib/watir/elements/element.rb:640:in `element_call'
	from /var/lib/gems/2.3.0/gems/watir-6.2.0/lib/watir/user_editable.rb:11:in `set'
	from (irb):19
	from /var/lib/gems/2.3.0/gems/railties-4.2.0/lib/rails/commands/console.rb:110:in `start'
	from /var/lib/gems/2.3.0/gems/railties-4.2.0/lib/rails/commands/console.rb:9:in `start'
	from /var/lib/gems/2.3.0/gems/railties-4.2.0/lib/rails/commands/commands_tasks.rb:68:in `console'
	from /var/lib/gems/2.3.0/gems/railties-4.2.0/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
	from /var/lib/gems/2.3.0/gems/railties-4.2.0/lib/rails/commands.rb:17:in `<top (required)>'
	from script/rails:6:in `require'
	from script/rails:6:in `<main>'
```

In order to help us efficiently investigate your issue, please provide the following information:

## Firefox Version
From the "About" dialog.

## Platform
e.g. Windows/OSX/Linux

## Steps to reproduce -
Please provide a description of how to reproduce the problem, including the following essential elements:

 - [ ] A *minimal* HTML file (or files) that can be used to reproduce the issue (this should not be your production site, unless you are unable to isolate the problem, but a page containing only the elements required to observe the problem).

 - [ ] A [trace level log](https://github.com/mozilla/geckodriver#firefox-capabilities) from a geckodriver session that demonstrates the problem using the minimal example file. Note: this is *not* the same as a selenium log.

** If any of the above are missing we will have to unfortunately close your issue. We will gladly reopen the issue once all the information requested has been added **

**Comment by Xetoris** [comment]
I'm seeing this too. 

Details:
Ruby
FF v 53
GeckoDriver V 16.1
Selenium Gem V 3.4

Exception happens when you try to .set which I believe boils down to send_keys. I attempted with just a basic html page with a single input and got the exception.

Unfortunately I can't seem to find an easy way to set the log level for the geckodriver to capture that output.

**Comment by teknologist** [comment]
Same here with GeckoDriver V 16.1
Firefox 53.0
and selenium 6.4.1

**Comment by andreastt** [comment]
@khanimdad Your output says you’re using `selenium-webdriver-3.3.0`, that is version 3.3.0 of the selenium-webdriver Ruby gem. You need the latest version, which at the time of writing is 3.4.0.

Judging by https://github.com/SeleniumHQ/selenium/blame/master/rb/lib/selenium/webdriver/remote/w3c/bridge.rb#L365, it looks like the Ruby implementation is using the correct `text` field to pass a string to geckodriver, but I can’t tell exactly why any of you are having this issue because you did not attach a trace-level geckodriver log.

In any case, I suspect that this issue is a Selenium issue as we have seen similar reports about other language bindings when we made the switch for the Element Send Keys command to only accept `text`.

I am happy to re-open the issue if you think it is still a geckodriver issue and you attach a trace-level log.

**Comment by teknologist** [comment]
@andreastt  Thanks for looking into this.
I my case I am using javascript and my npm versions are up to date:

"selenium-standalone": "^6.4.1",
"geckodriver": "^1.6.1",

With FF 53 it fails with the exact same error when setValue of an input field.
When I rollback to FF 52 it works like it used to.

IMHO I don't think this has anything to do with ruby, it is related to upgrading to latest FF version, that is 53.0

**Comment by andreastt** [comment]
I don’t know what selenium-standalone or geckodriver is in JS, but those are not packages we maintain. This issue is also about the Ruby bindings failing to send keys; not about Node.js.

When I said earlier that “your mileage may vary”, I meant exactly this. I don’t think this problem as been fixed in the Node.js bindings that Selenium provides yet. Please report a bug with them.

> IMHO I don't think this has anything to do with ruby, it is related to upgrading to latest FF version, that is 53.0 (As you stated)

It does. Let me explain.

geckodriver’s API changed with the 0.16.0 release to take `{"text": "foobar"}` as the message structure for the POST `/session/{session id}/element/{element id}/value` endpoint, known as the [Element Send Keys](https://w3c.github.io/webdriver/webdriver-spec.html#element-send-keys) command.

Firefox [changed around the same time](https://bugzilla.mozilla.org/show_bug.cgi?id=1348782) to only accept a `text` field. This change was since uplifted to Firefox 53, and is consequently not part of Firefox 52.

The error message `Expected [object Undefined] undefined to be a string` originates from the Marionette remote protocol, which sits inside Firefox. This means that geckodriver passed the wrong message format. Probably this is because you’re using a geckodriver version that is too old, a broken Selenium language binding, an supported Firefox version, or a combination of the above.

**Comment by teknologist** [comment]
@andreastt  Thanks for the update and taking time to explain.

**Comment by wallheater** [comment]
Resolved for me (on Mac) by `brew upgrade geckodriver` to get version **0.16.0**.

**Comment by andreastt** [comment]
For the record, the latest available version of geckodriver is 0.16.1, and the Homebrew package is not supported by Mozilla. Use at your own risk.
